base/compiler/inferencestate.jl

```diff
@@ -1129,24 +1129,36 @@ end
 """
     Future{T}
 
-Delayed return value for a value of type `T`, similar to RefValue{T}, but
-explicitly represents completed as a `Bool` rather than as `isdefined`.
-Set once with `f[] = v` and accessed with `f[]` afterwards.
+Assign-once delayed return value for a value of type `T`, similar to RefValue{T}.
+Can be constructed in one of three ways:
 
-Can also be constructed with the `completed` flag value and a closure to
-produce `x`, as well as the additional arguments to avoid always capturing the
-same couple of values.
+1. With an immediate as `Future{T}(val)`
+2. As an assign-once storage location with `Future{T}()`. Assigned (once) using `f[] = val`.
+3. As a delayed computation with `Future{T}(callback, dep, interp, sv)` to have
+   `sv` arrange to call the `callback` with the result of `dep` when it is ready.
+
+Use `isready` to check if the value is ready, and `getindex` to get the value.
+
+In general, `Future` is expected to be short-lived and have its allocation elided.
```

Comment:
This isn't true. There is no expectations on the lifetime if it is captured for later

Reply:
That's why I said `in general`. The implementation only captures `later` not the future itself, so the allocation is still elided. I can remove it from the docstring though, since it's a bit of an implementation detail.